demos/bubble_shear-goal.py

```diff
@@ -0,0 +1,313 @@
+
+
+def get_bcs(mesh_seq):
+    def bcs(index):
+        return [DirichletBC(mesh_seq.function_spaces["c"][index], 0.0, "on_boundary")]
+
+    return bcs
+
+
+def ball_initial_condition(x, y):
+    ball_r0, ball_x0, ball_y0 = 0.15, 0.5, 0.65
+    r = sqrt(pow(x - ball_x0, 2) + pow(y - ball_y0, 2))
+    return conditional(r < ball_r0, 1.0, 0.0)
+
+
+def get_initial_condition(mesh_seq):
+    fs = mesh_seq.function_spaces["c"][0]
+    x, y = SpatialCoordinate(mesh_seq[0])
+    ball = ball_initial_condition(x, y)
+    c0 = assemble(interpolate(ball, fs))
+    return {"c": c0}
+
+
+def get_solver(mesh_seq):
+    def solver(index, ic):
+        tp = mesh_seq.time_partition
+        t_start, t_end = tp.subintervals[index]
+        dt = tp.timesteps[index]
+
+        # Initialise the concentration fields
+        Q = mesh_seq.function_spaces["c"][index]
+        c = Function(Q, name="c")
+        c_ = Function(Q, name="c_old").assign(ic["c"])
+
+        # Specify the velocity function space
+        V = VectorFunctionSpace(mesh_seq[index], "CG", 1)
+        u = Function(V)
+        u_ = Function(V)
+
+        # Compute the velocity field at t_start and assign it to u_
+        x, y = SpatialCoordinate(mesh_seq[index])
+        u_.interpolate(velocity_expression(x, y, t_start))
+
+        # We pass both the concentration and velocity Functions to get_form
+        form_fields = {"c": (c, c_), "u": (u, u_)}
+        F = mesh_seq.form(index, form_fields)["c"]
+        nlvp = NonlinearVariationalProblem(F, c, bcs=mesh_seq.bcs(index))
+        nlvs = NonlinearVariationalSolver(nlvp, ad_block_tag="c")
+
+        # Time integrate from t_start to t_end
+        t = t_start + dt
+        while t < t_end + 0.5 * dt:
+            # update the background velocity field at the current timestep
+            u.interpolate(velocity_expression(x, y, t))
+
+            # solve the advection equation
+            nlvs.solve()
+
+            # update the 'lagged' concentration and velocity field
+            c_.assign(c)
+            u_.assign(u)
+            t += dt
+
+        return {"c": c}
+
+    return solver
+
+
+# In the previous demo, the :meth:`get_form` method was only called from within the
+# :meth:`get_solver`, so we were able to pass the velocity field as an argument when
+# calling :meth:`get_form`. However, in the context of goal-oriented mesh adaptation,
+# the :meth:`get_form` method is called from within :class:`GoalOrientedMeshSeq`
+# while computing error indicators. There, only those fields that are passed to
+# :class:`GoalOrientedMeshSeq` are passed to :meth:`get_form`. This means that the velocity
+# field would not be updated in time. In such a case, we will manually compute the
+# velocity field using the current simulation time, which will be passed automatically
+# as an :arg:`err_ind_time` kwarg in :meth:`GoalOrientedMeshSeq.indicate_errors`. ::
+
+
+def get_form(mesh_seq):
+    def form(index, form_fields, err_ind_time=None):
```

Comment:
I wonder if we should just add `time` as an argument to `form` in general? For problems where the form doesn't vary with time, it will just be ignored. IMO it would be better to only pass solution fields (and their lagged counterparts) through the second argument and always determine other variables based on the current time. I guess this would imply some refactoring, including the other demos.

It seems odd to reference error indicators in the form expression.

Reply:
If you do end up making this change then it would be handy to fix #140 while you're at it if you don't mind.

Reply:
Sorry @acse-ej321, I just saw this! Please see the new comment I left explaining my reasoning here. I agree that this suggestion is much nicer but I think it's also much more inefficient - unless I'm missing something! Please correct me if I'm wrong :)